Add option to exclude addons in Quick Open dialog

hilfazer opened the pull request:
One of the features requested in https://github.com/godotengine/godot-proposals/issues/25.
A video with the feature working in 3.3:

https://user-images.githubusercontent.com/29497869/113396134-49110c00-939b-11eb-8230-f7e4698d2ab6.mp4



<!--
Pull requests should always be made for the `master` branch first, as that's
where development happens and the source of all future stable release branches.

Relevant fixes are cherry-picked for stable branches as needed.

Do not create a pull request for stable branches unless the change is already
available in the `master` branch and it cannot be easily cherry-picked.
Alternatively, if the change is only relevant for that branch (e.g. rendering
fixes for the 3.2 branch).
-->

hinlopen commented on editor/quick_open.cpp:
```diff
@@ -224,13 +232,23 @@ void EditorQuickOpen::_notification(int p_what) {
 	search_box->set_right_icon(search_options->get_theme_icon("Search", "EditorIcons"));
 }
 
+void EditorQuickOpen::_save_exclude_addons() {
```
Why is this a project setting but also a checkbox in the dialog? I am against introducing checkboxes to Quick Open dialogs, but a project setting could be fine. That would allow this filter to be done just once when opening the dialog in `_build_search_cache`.

hilfazer replied:
Imagine the user doesn't want to see addon files in his/her current project. But there came a time when s/he needs to search for a file from addons (for whatever reason). S/he opens project settings, find the setting, clicks it, opens the Quick Open dialog, finds the file s/he needs, opens the project settings again, finds the setting and clicks it again.
Unless there's a checkbox for than in Quick Open dialog that simplifies that process a lot. Which is the reason i've added it.

Calinou replied:
I agree, it makes sense to expose this in a checkbox in the Quick Open dialogs. Its value should persist across restarts.

hilfazer replied:
I made it a project setting since the user might have some project when s/he doesn't want to see addons' files and some projects (like addons s/he's developing) where s/he does. But if editor settings is a better place for that, sure, i can change.

hinlopen replied:
@hilfazer I don't imagine Quick Open is the first thing they will use to find it, I imagine they will just use the filesystem dock. But even this unlikely scenario you describe doesn't sound unreasonable. It's certainly preferable over introducing edge case checkboxes in every place, especially Quick Open dialogs. Unnecessary elements like this are a distraction that go against its purpose.

hilfazer commented:
I've replaced project setting with an editor setting. I'm not sure if "filesystem/file_dialog/quick_open_exclude_addons" is a right way to name it so i'm asking for some guidance here.

Hinlopen made a good point about using the filesystem dock so i've removed the checkbox.

Calinou reviewed (approved):
Tested locally (rebased against `master` in https://github.com/Calinou/godot/tree/quick_open_exclude_addons_40), it works as expected.

### Setting disabled (default)

![Screenshot_20230802_222341](https://github.com/godotengine/godot/assets/180032/2d47c3ad-c05a-41e4-bf84-adf4ba65e7e9)

### Setting enabled

![Screenshot_20230802_222353](https://github.com/godotengine/godot/assets/180032/f30898a0-6073-47f2-ac7d-d28048b20778)

Code looks good to me.

YuriSizov commented:
I feel like this option is a bit too specific. Are we going to add a new boolean for each case where we may want to exclude the addons folder? And the other way around, what if there is a use case for ignoring more folders for "Quick Open"?

hilfazer commented:
> I feel like this option is a bit too specific. Are we going to add a new boolean for each case where we may want to exclude the addons folder? And the other way around, what if there is a use case for ignoring more folders for "Quick Open"?

In case there should be a list of folders to ignore should it still be an editor setting or rather a project setting? While addons/ is universal across all projects the others folders a user might want to ignore will/can differ from project to project.

akien-mga commented:
I think this kind of filter should be toggled directly in the Quick Open dialog, and not in editor settings.

So if we think we only want to support excluding results from `res://addons`, we can add a button for this in that dialog.

If we want to support more arbitrary exclude filters, then we could add a way to hide folders with a right click tooltip, and add a button to toggle showing hidden folders. `res://addons` could be hidden by default, and more folders could be hidden on a per project basis (saved in project metadata), and unhidden via showing hidden folders (and thus `res://addons` could be unhidden too when desired). But I don't know if we need that level of flexibility.

YuriSizov commented:
There are already requests to exclude addons from other places as well, e.g. https://github.com/godotengine/godot-proposals/issues/7392 for the FileSystem dock filter. Which is why I'm raising my concern.

AThousandShips commented on editor/editor_quick_open.cpp:
```diff
@@ -75,7 +80,9 @@ void EditorQuickOpen::_build_search_cache(EditorFileSystemDirectory *p_efsd) {
 		for (String &parent_type : base_types) {
 			if (ClassDB::is_parent_class(engine_type, parent_type) || EditorNode::get_editor_data().script_class_is_parent(script_type, parent_type)) {
-				files.push_back(file.substr(6, file.length()));
+				if (!(skip_addons && file.substr(6, file.length()).begins_with(addons_directory + "/"))) {
```
```suggestion
				if (!(skip_addons && file.substr(6).begins_with(addons_directory + "/"))) {
```
The length here makes it take 6 characters beyond the end, though it caps out so it just takes the full string, which is equivalent to this, alternatively:
```suggestion
				if (!(skip_addons && file.right(-6).begins_with(addons_directory + "/"))) {
```

AThousandShips commented on editor/editor_quick_open.cpp:
```diff
@@ -75,7 +80,9 @@ void EditorQuickOpen::_build_search_cache(EditorFileSystemDirectory *p_efsd) {
 			if (ClassDB::is_parent_class(engine_type, parent_type) || EditorNode::get_editor_data().script_class_is_parent(script_type, parent_type)) {
-				files.push_back(file.substr(6, file.length()));
+				if (!(skip_addons && file.substr(6, file.length()).begins_with(addons_directory + "/"))) {
+					files.push_back(file.substr(6, file.length()));
```
```suggestion
					files.push_back(file.substr(6));
```
Same here

AThousandShips commented on editor/editor_quick_open.cpp:
```diff
@@ -33,6 +33,9 @@
 #include "core/os/keyboard.h"
 #include "editor/editor_node.h"
 #include "editor/editor_scale.h"
+#include "editor_settings.h"
```
```suggestion
#include "editor/editor_settings.h"
```
The rest here have the full path, so use here as well

AThousandShips reviewed (approved):
Some minor details, looks good otherwise

AThousandShips commented on editor/editor_quick_open.cpp:
```diff
@@ -75,7 +80,9 @@ void EditorQuickOpen::_build_search_cache(EditorFileSystemDirectory *p_efsd) {
 		for (String &parent_type : base_types) {
 			if (ClassDB::is_parent_class(engine_type, parent_type) || EditorNode::get_editor_data().script_class_is_parent(script_type, parent_type)) {
-				files.push_back(file.substr(6, file.length()));
+				if (!(skip_addons && file.substr(6).begins_with(addons_directory + "/"))) {
```
```suggestion
				if (!skip_addons || !file.substr(6).begins_with(addons_directory + "/")) {
```
Simpler and easier to read than the grouped form IMO

AThousandShips commented:
Also needs a rebase against the most recent version, it has conflicts

hilfazer committed:
Add an option to exclude addons in Quick Open dialog

garf commented:
Why not to just allow to exclude folders right-clicking them? Something like that in context menu: 

Mark Directory as... > 
    Excluded from Search
    Excluded from Quick Search
    Ignored

This might be scalable and also you can stay in context.

cyda89 commented:
> Why not to just allow to exclude folders right-clicking them? Something like that in context menu:
> 
> ```
> Mark Directory as... > 
>     Excluded from Search
>     Excluded from Quick Search
>     Ignored
> ```
> 
> This might be scalable and also you can stay in context and show the states directly on the folder

This is excactly what i had in mind, because that is how it works in other IDEs (e.g. IntelliJ Idea). I would also add an option to exclude from "Find in Files" since i think it would also greatly benefit from this change.

fire commented:
I added a salvageable tag because there seems to be interest in having this feature.

arkology commented:
As I see this is implemented in https://github.com/godotengine/godot/pull/56772
cc @Repiteo

akien-mga commented:
Superseded by #56772. Thanks for the contribution!
